redis/src/main/java/org/hibernate/ogm/datastore/redis/RedisHashDialect.java

```diff
@@ -0,0 +1,282 @@
+import org.hibernate.ogm.model.key.spi.AssociationKeyMetadata;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.key.spi.EntityKeyMetadata;
+import org.hibernate.ogm.model.key.spi.RowKey;
+import org.hibernate.ogm.model.spi.Association;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.hibernate.ogm.model.spi.TupleOperation;
+import org.hibernate.ogm.type.spi.GridType;
+import org.hibernate.type.Type;
+
+import com.lambdaworks.redis.KeyScanCursor;
+import com.lambdaworks.redis.ScanArgs;
+
+/**
+* Stores tuples and associations inside Redis using hash data structures.
+ * <p>
+ * Tuples are stored in Redis within hashes. Associations are stored in Redis obtained as either single values or a
+ * JSON serialization of a {@link org.hibernate.ogm.datastore.redis.dialect.value.Association} object in list data structures.
+ *
+ * @author Mark Paluch
+ */
+public class RedisHashDialect extends AbstractRedisDialect {
+
+	public RedisHashDialect(RedisDatastoreProvider provider) {
+		super( provider.getConnection() );
+	}
+
+	@Override
+	public GridType overrideType(Type type) {
+		return RedisHashTypeConverter.INSTANCE.convert( type );
+	}
+
+	@Override
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	public Tuple getTuple(
+			EntityKey key, TupleContext tupleContext) {
+		String entityIdString = entityId( key );
+		if ( !connection.exists( entityIdString ) ) {
+			return null;
+		}
+
+		Map<String, Object> objects;
+		if ( tupleContext.getSelectableColumns().isEmpty() ) {
+			objects = (Map) connection.hgetall( entityIdString );
+		}
+		else {
+			List<String> hmget = connection.hmget( entityIdString, getFields( tupleContext ) );
+			objects = toEntity( tupleContext, hmget );
+		}
+
+		return new Tuple( new RedisTupleSnapshot( objects ) );
+	}
+
+	private Map<String, Object> toEntity(TupleContext tupleContext, List<String> hmget) {
+		Map<String, Object> objects = new HashMap<>();
+		for ( int i = 0; i < tupleContext.getSelectableColumns().size(); i++ ) {
+			String columnName = tupleContext.getSelectableColumns().get( i );
+			String value = hmget.get( i );
+			if ( value == null ) {
+				continue;
+			}
+			objects.put( columnName, value );
+		}
+		return objects;
+	}
+
+	private String[] getFields(TupleContext tupleContext) {
+		return tupleContext.getSelectableColumns().toArray( new String[tupleContext.getSelectableColumns().size()] );
+	}
+
+	@Override
+	public void insertOrUpdateTuple(
+			EntityKey key, Tuple tuple, TupleContext tupleContext) throws TupleAlreadyExistsException {
+		Map<String, Object> map = ( (RedisTupleSnapshot) tuple.getSnapshot() ).getMap();
+		Map<String, String> entity = new HashMap<>();
+		List<String> toDelete1 = new ArrayList<>();
+
+		for ( TupleOperation action : tuple.getOperations() ) {
+			switch ( action.getType() ) {
+				case PUT:
+					map.put( action.getColumn(), action.getValue() );
+
+					// TODO: IllegalStateException when value is not String/Character?
```

Comment:
Yep. The `NumericBoolean` or the `EnumType`  (or any other custom types) can produce types other than `String`. This place would be a good fit to catch those values and raise a better exception than `ClassCastException` in a later stage. WDYT?